lightning/src/routing/onion_message.rs

```diff
@@ -0,0 +1,283 @@
+		let (secp_ctx, network_graph, _, _, logger) = build_line_graph(features);
+		let (_, our_id, _, node_pks) = get_nodes(&secp_ctx);
+
+		let path = super::find_path(&our_id, &node_pks[18], &network_graph, None, Arc::clone(&logger)).unwrap();
+		assert_eq!(path.len(), 19);
+	}
+
+	#[test]
+	fn disable_nodes_test() {
+		// Check that we won't route over nodes that require unknown feature bits.
+		let mut features = InitFeatures::empty();
+		features.set_onion_messages_optional();
+		let (secp_ctx, network_graph, gossip_sync, _, logger) = build_graph_with_features(features.to_context());
+		let (_, our_id, privkeys, node_pks) = get_nodes(&secp_ctx);
+
+		// Disable nodes 1, 2, and 8 by requiring unknown feature bits
+		let mut unknown_features = NodeFeatures::empty();
+		unknown_features.set_unknown_feature_required();
+		add_or_update_node(&gossip_sync, &secp_ctx, &privkeys[0], unknown_features.clone(), 1);
+		add_or_update_node(&gossip_sync, &secp_ctx, &privkeys[1], unknown_features.clone(), 1);
+		add_or_update_node(&gossip_sync, &secp_ctx, &privkeys[7], unknown_features.clone(), 1);
+
+		// If all nodes require some features we don't understand, route should fail
+		let err = super::find_path(&our_id, &node_pks[2], &network_graph, None, Arc::clone(&logger)).unwrap_err();
+		assert_eq!(err, super::Error::PathNotFound);
+
+		// If we specify a channel to node7, that overrides our local channel view and that gets used
+		let our_chans = vec![get_channel_details(Some(42), node_pks[7].clone(), features, 250_000_000)];
+		let path = super::find_path(&our_id, &node_pks[2], &network_graph, Some(&our_chans.iter().collect::<Vec<_>>()), Arc::clone(&logger)).unwrap();
+		assert_eq!(path.len(), 2);
+		assert_eq!(path[0], node_pks[7]);
+		assert_eq!(path[1], node_pks[2]);
+	}
+
+	#[test]
+	fn disabled_channels_test() {
+		// Check that we won't attempt to route over nodes where the channel is disabled from their
+		// direction (implying the peer is offline).
+		let mut features = InitFeatures::empty();
+		features.set_onion_messages_optional();
+		let (secp_ctx, network_graph, _, _, logger) = build_graph_with_features(features.to_context());
+		let (_, our_id, _, node_pks) = get_nodes(&secp_ctx);
+
+		// Route to 1 via 2 and 3 because our channel to 1 is disabled
+		let path = super::find_path(&our_id, &node_pks[0], &network_graph, None, Arc::clone(&logger)).unwrap();
+		assert_eq!(path.len(), 3);
+		assert_eq!(path[0], node_pks[1]);
+		assert_eq!(path[1], node_pks[2]);
+		assert_eq!(path[2], node_pks[0]);
+
+		// If we specify a channel to node1, that overrides our local channel view and that gets used
+		let our_chans = vec![get_channel_details(Some(42), node_pks[0].clone(), features, 250_000_000)];
+		let path = super::find_path(&our_id, &node_pks[0], &network_graph, Some(&our_chans.iter().collect::<Vec<_>>()), Arc::clone(&logger)).unwrap();
+		assert_eq!(path.len(), 1);
+		assert_eq!(path[0], node_pks[0]);
+	}
+}
```

Comment:
Maybe test cases for a) checking that valid first hops are included and b) that invalid first hops are not considered (cf. `router::tests::invalid_first_hop_test`) might be nice.